File: DataEngineering/04_Batch_Processing/projects/golang/02_log_anomaly_detector/main.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"regexp"
	"sync"
)

func main() {
	inPath := flag.String("in", "server.log", "Input log file")
	outPath := flag.String("out", "anomalies.csv", "Output anomalies file")
	threshold := flag.Int("threshold", 10, "Threshold for 404 errors")
	workers := flag.Int("workers", 4, "Number of concurrent parser workers")
	flag.Parse()

	file, err := os.Open(*inPath)
	if err != nil {
		log.Printf("Failed to open %s: %v. Please provide a valid input.", *inPath, err)
		return
	}
	defer file.Close()

	// Pattern: 192.168.1.1 - - [10/Oct/2023:13:55:36 -0700] "GET /index.html HTTP/1.1" 404 2326
	logPattern := regexp.MustCompile(`^(\d{1,3}(?:\.\d{1,3}){3}).*?"\s+(\d{3})\s+`)

	lines := make(chan string, 1000)
	type result struct {
		ip     string
		status string
	}
	results := make(chan result, 1000)

	var wg sync.WaitGroup

	// Start parser workers
	for i := 0; i < *workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for line := range lines {
				matches := logPattern.FindStringSubmatch(line)
				if len(matches) == 3 {
					results <- result{ip: matches[1], status: matches[2]}
				}
			}
		}()
	}

	// Read lines
	go func() {
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			lines <- scanner.Text()
		}
		if err := scanner.Err(); err != nil {
			log.Printf("error reading file: %v", err)
		}
		close(lines)
	}()

	// Wait for workers to finish
	go func() {
		wg.Wait()
		close(results)
	}()

	// Aggregate errors
	errorCounts := make(map[string]int)
	for res := range results {
		if res.status == "404" {
			errorCounts[res.ip]++
		}
	}

	// Write output
	outFile, err := os.Create(*outPath)
	if err != nil {
		log.Fatalf("failed to create output: %v", err)
	}
	defer outFile.Close()

	outFile.WriteString("ip,error_count\n")
	anomaliesFound := 0
	for ip, count := range errorCounts {
		if count > *threshold {
			outFile.WriteString(fmt.Sprintf("%s,%d\n", ip, count))
			anomaliesFound++
		}
	}

	log.Printf("Anomaly detection complete. Found %d anomalies.", anomaliesFound)
}

File: DataEngineering/04_Batch_Processing/projects/python/02_log_anomaly_detector/main.py
import argparse
import logging
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, regexp_extract, count, when

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def main(input_path: str, output_path: str, threshold: int):
    """
    Detect anomalies in logs based on high 404 error rates per IP.
    """
    logger.info("Initializing SparkSession...")
    spark = SparkSession.builder \
        .appName("LogAnomalyDetector") \
        .getOrCreate()
        
    logger.info(f"Reading raw log lines from {input_path}...")
    try:
        # Read raw text lines
        raw_logs = spark.read.text(input_path)
    except Exception as e:
        logger.error(f"Failed to read input data: {e}")
        spark.stop()
        return

    # Extract IP and HTTP status code using regex
    # Common Apache/Nginx format: 192.168.1.1 - - [10/Oct/... "GET /..." 404 232
    ip_pattern = r'^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'
    status_pattern = r'"\s+(\d{3})\s+'

    df = raw_logs.select(
        regexp_extract(col("value"), ip_pattern, 1).alias("ip"),
        regexp_extract(col("value"), status_pattern, 1).alias("status")
    )

    # Filter out unparsed lines (where IP or status could not be extracted)
    df = df.filter((col("ip") != "") & (col("status") != ""))

    logger.info("Aggregating 404 errors by IP...")
    # Count occurrences of 404 status per IP
    error_counts = df.filter(col("status") == "404") \
        .groupBy("ip") \
        .agg(count("*").alias("error_count"))

    logger.info(f"Filtering IPs with error count > {threshold}...")
    anomalous_ips = error_counts.filter(col("error_count") > threshold)

    logger.info(f"Writing anomalous IPs to {output_path}...")
    try:
        anomalous_ips.write \
            .mode("overwrite") \
            .csv(output_path, header=True)
        logger.info("Anomaly detection completed successfully.")
    except Exception as e:
        logger.error(f"Failed to write output data: {e}")
        
    spark.stop()

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Log Anomaly Detector")
    parser.add_argument("--input-path", required=True, help="S3 URI for input raw logs (Text)")
    parser.add_argument("--output-path", required=True, help="S3 URI for anomalous IPs output (CSV)")
    parser.add_argument("--threshold", type=int, default=10, help="Error count threshold to flag as anomaly")
    args = parser.parse_args()
    
    main(args.input_path, args.output_path, args.threshold)
